File: src/cachibot/models/work.py
"""Work management models: Functions, Schedules, Work, Tasks, Jobs, Todos."""

from datetime import datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel, Field


# =============================================================================
# ENUMS
# =============================================================================


class WorkStatus(str, Enum):
    """Work execution status."""

    PENDING = "pending"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"
    PAUSED = "paused"


class TaskStatus(str, Enum):
    """Task execution status."""

    PENDING = "pending"
    READY = "ready"  # All dependencies met, can run
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    FAILED = "failed"
    BLOCKED = "blocked"  # Waiting on dependencies
    SKIPPED = "skipped"


class JobStatus(str, Enum):
    """Job execution status."""

    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class TodoStatus(str, Enum):
    """Todo status."""

    OPEN = "open"
    DONE = "done"
    DISMISSED = "dismissed"


class Priority(str, Enum):
    """Priority levels."""

    LOW = "low"
    NORMAL = "normal"
    HIGH = "high"
    URGENT = "urgent"


class ScheduleType(str, Enum):
    """Schedule trigger types."""

    CRON = "cron"  # Cron expression
    INTERVAL = "interval"  # Every N seconds/minutes/hours
    ONCE = "once"  # One-time scheduled execution
    EVENT = "event"  # Triggered by an event


class FailureAction(str, Enum):
    """What to do when a step fails."""

    STOP = "stop"
    CONTINUE = "continue"
    SKIP_DEPENDENTS = "skip_dependents"


# =============================================================================
# FUNCTION (Reusable Template/Procedure)
# =============================================================================


class FunctionStep(BaseModel):
    """A step within a function template."""

    order: int = Field(description="Execution order (1-based)")
    name: str = Field(description="Step name")
    description: str | None = Field(default=None)
    action: str = Field(description="What to do (natural language or tool call)")
    depends_on: list[int] = Field(
        default_factory=list, description="Step orders this depends on"
    )
    retry_count: int = Field(default=0, description="Number of retries on failure")
    timeout_seconds: int | None = Field(default=None, description="Step timeout")
    on_failure: FailureAction = Field(
        default=FailureAction.STOP, description="What to do on failure"
    )


class FunctionParameter(BaseModel):
    """Parameter definition for a function."""

    name: str = Field(description="Parameter name")
    type: str = Field(default="string", description="Parameter type")
    default: Any | None = Field(default=None, description="Default value")
    required: bool = Field(default=True, description="Whether required")
    description: str | None = Field(default=None)


class BotFunction(BaseModel):
    """A reusable procedure/template that can be instantiated as Work."""

    id: str = Field(description="Unique function ID")
    bot_id: str = Field(description="Bot that owns this function")
    name: str = Field(description="Function name")
    description: str | None = Field(default=None)
    version: str = Field(default="1.0.0")

    # Steps
    steps: list[FunctionStep] = Field(default_factory=list)

    # Parameters (inputs when instantiating)
    parameters: list[FunctionParameter] = Field(default_factory=list)

    # Metadata
    tags: list[str] = Field(default_factory=list)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

    # Stats
    run_count: int = Field(default=0)
    last_run_at: datetime | None = Field(default=None)
    success_rate: float = Field(default=0.0)


# =============================================================================
# SCHEDULE (Cron/Timer)
# =============================================================================


class Schedule(BaseModel):
    """A schedule for triggering Work."""

    id: str = Field(description="Unique schedule ID")
    bot_id: str = Field(description="Bot that owns this schedule")
    name: str = Field(description="Schedule name")
    description: str | None = Field(default=None)

    # What to run
    function_id: str | None = Field(
        default=None, description="Function to instantiate"
    )
    function_params: dict[str, Any] = Field(
        default_factory=dict, description="Parameters for function"
    )

    # When to run
    schedule_type: ScheduleType = Field(default=ScheduleType.CRON)
    cron_expression: str | None = Field(
        default=None, description="Cron expression (if type=cron)"
    )
    interval_seconds: int | None = Field(
        default=None, description="Interval in seconds (if type=interval)"
    )
    run_at: datetime | None = Field(
        default=None, description="One-time run datetime (if type=once)"
    )
    event_trigger: str | None = Field(
        default=None, description="Event name (if type=event)"
    )
    timezone: str = Field(default="UTC")

    # Constraints
    enabled: bool = Field(default=True)
    max_concurrent: int = Field(default=1, description="Max concurrent Work instances")
    catch_up: bool = Field(
        default=False, description="Run missed executions on startup"
    )

    # Metadata
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    next_run_at: datetime | None = Field(default=None)
    last_run_at: datetime | None = Field(default=None)
    run_count: int = Field(default=0)


# =============================================================================
# WORK (Objective/Goal)
# =============================================================================


class Work(BaseModel):
    """A high-level objective the bot needs to accomplish."""

    id: str = Field(description="Unique work ID")
    bot_id: str = Field(description="Bot assigned to this work")
    chat_id: str | None = Field(default=None, description="Related chat context")

    # Identity
    title: str = Field(description="Work title")
    description: str | None = Field(default=None)
    goal: str | None = Field(
        default=None, description="Success criteria / definition of done"
    )

    # Source
    function_id: str | None = Field(
        default=None, description="Source function template"
    )
    schedule_id: str | None = Field(
        default=None, description="Source schedule (if recurring)"
    )
    parent_work_id: str | None = Field(
        default=None, description="Parent work (for sub-work)"
    )

    # Status
    status: WorkStatus = Field(default=WorkStatus.PENDING)
    priority: Priority = Field(default=Priority.NORMAL)
    progress: float = Field(
        default=0.0, ge=0.0, le=1.0, description="0.0-1.0 derived from tasks"
    )

    # Timing
    created_at: datetime = Field(default_factory=datetime.utcnow)
    started_at: datetime | None = Field(default=None)
    completed_at: datetime | None = Field(default=None)
    due_at: datetime | None = Field(default=None)

    # Results
    result: Any | None = Field(default=None)
    error: str | None = Field(default=None)

    # Context (parameters passed when instantiating)
    context: dict[str, Any] = Field(default_factory=dict)

    # Metadata
    tags: list[str] = Field(default_factory=list)


# =============================================================================
# TASK (Step within Work)
# =============================================================================


class Task(BaseModel):
    """A discrete, completable step within Work."""

    id: str = Field(description="Unique task ID")
    bot_id: str = Field(description="Bot assigned to this task")
    work_id: str = Field(description="Parent work")
    chat_id: str | None = Field(default=None)

    # Identity
    title: str = Field(description="Task title")
    description: str | None = Field(default=None)
    action: str | None = Field(default=None, description="Specific action to perform")

    # Ordering & Dependencies
    order: int = Field(default=0, description="Execution order within work")
    depends_on: list[str] = Field(
        default_factory=list, description="Task IDs this depends on"
    )

    # Status
    status: TaskStatus = Field(default=TaskStatus.PENDING)
    priority: Priority = Field(default=Priority.NORMAL)

    # Execution config
    retry_count: int = Field(default=0)
    max_retries: int = Field(default=3)
    timeout_seconds: int | None = Field(default=None)

    # Timing
    created_at: datetime = Field(default_factory=datetime.utcnow)
    started_at: datetime | None = Field(default=None)
    completed_at: datetime | None = Field(default=None)

    # Results
    result: Any | None = Field(default=None)
    error: str | None = Field(default=None)


# =============================================================================
# JOB (Execution attempt of Task)
# =============================================================================


class Job(BaseModel):
    """A single execution attempt of a Task."""

    id: str = Field(description="Unique job ID")
    bot_id: str = Field(description="Bot running this job")
    task_id: str = Field(description="Task being executed")
    work_id: str = Field(description="Parent work")
    chat_id: str | None = Field(default=None)

    # Status
    status: JobStatus = Field(default=JobStatus.PENDING)
    attempt: int = Field(default=1, description="Attempt number (1, 2, 3...)")
    progress: float = Field(default=0.0, ge=0.0, le=1.0)

    # Timing
    created_at: datetime = Field(default_factory=datetime.utcnow)
    started_at: datetime | None = Field(default=None)
    completed_at: datetime | None = Field(default=None)

    # Results
    result: Any | None = Field(default=None)
    error: str | None = Field(default=None)

    # Logs
    logs: list[dict] = Field(default_factory=list, description="Execution logs")


class JobLog(BaseModel):
    """A log entry for a job."""

    timestamp: datetime = Field(default_factory=datetime.utcnow)
    level: str = Field(default="info", description="debug, info, warn, error")
    message: str = Field(description="Log message")
    data: Any | None = Field(default=None, description="Additional data")


# =============================================================================
# TODO (Reminder/Note)
# =============================================================================


class Todo(BaseModel):
    """A lightweight reminder or note."""

    id: str = Field(description="Unique todo ID")
    bot_id: str = Field(description="Bot this todo belongs to")
    chat_id: str | None = Field(default=None, description="Related chat")

    # Content
    title: str = Field(description="Todo title")
    notes: str | None = Field(default=None)

    # Status
    status: TodoStatus = Field(default=TodoStatus.OPEN)
    priority: Priority = Field(default=Priority.NORMAL)

    # Timing
    created_at: datetime = Field(default_factory=datetime.utcnow)
    completed_at: datetime | None = Field(default=None)
    remind_at: datetime | None = Field(
        default=None, description="Optional reminder time"
    )

    # Conversion tracking
    converted_to_work_id: str | None = Field(default=None)
    converted_to_task_id: str | None = Field(default=None)

    # Metadata
    tags: list[str] = Field(default_factory=list)


# =============================================================================
# LIST RESPONSE MODELS
# =============================================================================


class WorkList(BaseModel):
    """List of work items."""

    items: list[Work] = Field(default_factory=list)
    total: int = Field(default=0)


class TaskList(BaseModel):
    """List of tasks."""

    items: list[Task] = Field(default_factory=list)
    total: int = Field(default=0)


class JobList(BaseModel):
    """List of jobs."""

    items: list[Job] = Field(default_factory=list)
    total: int = Field(default=0)


class TodoList(BaseModel):
    """List of todos."""

    items: list[Todo] = Field(default_factory=list)
    total: int = Field(default=0)


class FunctionList(BaseModel):
    """List of functions."""

    items: list[BotFunction] = Field(default_factory=list)
    total: int = Field(default=0)


class ScheduleList(BaseModel):
    """List of schedules."""

    items: list[Schedule] = Field(default_factory=list)
    total: int = Field(default=0)

File: mobile/lib/models/work.dart
enum WorkStatus {
  pending,
  inProgress,
  completed,
  failed,
  cancelled,
  paused;

  static WorkStatus fromString(String value) {
    switch (value) {
      case 'in_progress':
        return WorkStatus.inProgress;
      default:
        return WorkStatus.values.firstWhere(
          (e) => e.name == value,
          orElse: () => WorkStatus.pending,
        );
    }
  }

  String toApi() {
    switch (this) {
      case WorkStatus.inProgress:
        return 'in_progress';
      default:
        return name;
    }
  }
}

enum TaskStatus {
  pending,
  ready,
  inProgress,
  completed,
  failed,
  blocked,
  skipped;

  static TaskStatus fromString(String value) {
    switch (value) {
      case 'in_progress':
        return TaskStatus.inProgress;
      default:
        return TaskStatus.values.firstWhere(
          (e) => e.name == value,
          orElse: () => TaskStatus.pending,
        );
    }
  }

  String toApi() {
    switch (this) {
      case TaskStatus.inProgress:
        return 'in_progress';
      default:
        return name;
    }
  }
}

enum JobStatus {
  pending,
  running,
  completed,
  failed,
  cancelled;

  static JobStatus fromString(String value) {
    return JobStatus.values.firstWhere(
      (e) => e.name == value,
      orElse: () => JobStatus.pending,
    );
  }
}

enum TodoStatus {
  open,
  done,
  dismissed;

  static TodoStatus fromString(String value) {
    return TodoStatus.values.firstWhere(
      (e) => e.name == value,
      orElse: () => TodoStatus.open,
    );
  }
}

enum Priority {
  low,
  normal,
  high,
  urgent;

  static Priority fromString(String value) {
    return Priority.values.firstWhere(
      (e) => e.name == value,
      orElse: () => Priority.normal,
    );
  }
}

class BotWork {
  const BotWork({
    required this.id,
    required this.botId,
    this.chatId,
    required this.title,
    this.description,
    this.goal,
    this.functionId,
    this.scheduleId,
    this.parentWorkId,
    required this.status,
    required this.priority,
    required this.progress,
    required this.createdAt,
    this.startedAt,
    this.completedAt,
    this.dueAt,
    this.result,
    this.error,
    this.context,
    this.tags = const [],
    this.taskCount,
    this.completedTaskCount,
  });

  final String id;
  final String botId;
  final String? chatId;
  final String title;
  final String? description;
  final String? goal;
  final String? functionId;
  final String? scheduleId;
  final String? parentWorkId;
  final WorkStatus status;
  final Priority priority;
  final double progress;
  final DateTime createdAt;
  final DateTime? startedAt;
  final DateTime? completedAt;
  final DateTime? dueAt;
  final String? result;
  final String? error;
  final Map<String, dynamic>? context;
  final List<String> tags;
  final int? taskCount;
  final int? completedTaskCount;

  factory BotWork.fromJson(Map<String, dynamic> json) {
    return BotWork(
      id: json['id'] as String,
      botId: json['botId'] as String,
      chatId: json['chatId'] as String?,
      title: json['title'] as String,
      description: json['description'] as String?,
      goal: json['goal'] as String?,
      functionId: json['functionId'] as String?,
      scheduleId: json['scheduleId'] as String?,
      parentWorkId: json['parentWorkId'] as String?,
      status: WorkStatus.fromString(json['status'] as String),
      priority: Priority.fromString(json['priority'] as String? ?? 'normal'),
      progress: (json['progress'] as num?)?.toDouble() ?? 0.0,
      createdAt: DateTime.parse(json['createdAt'] as String),
      startedAt: json['startedAt'] != null
          ? DateTime.parse(json['startedAt'] as String)
          : null,
      completedAt: json['completedAt'] != null
          ? DateTime.parse(json['completedAt'] as String)
          : null,
      dueAt: json['dueAt'] != null
          ? DateTime.parse(json['dueAt'] as String)
          : null,
      result: json['result'] as String?,
      error: json['error'] as String?,
      context: json['context'] as Map<String, dynamic>?,
      tags: (json['tags'] as List<dynamic>?)?.cast<String>() ?? [],
      taskCount: json['taskCount'] as int?,
      completedTaskCount: json['completedTaskCount'] as int?,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'botId': botId,
        'chatId': chatId,
        'title': title,
        'description': description,
        'goal': goal,
        'status': status.toApi(),
        'priority': priority.name,
        'progress': progress,
        'createdAt': createdAt.toIso8601String(),
        'startedAt': startedAt?.toIso8601String(),
        'completedAt': completedAt?.toIso8601String(),
        'dueAt': dueAt?.toIso8601String(),
        'tags': tags,
      };

  BotWork copyWith({
    String? title,
    String? description,
    String? goal,
    WorkStatus? status,
    Priority? priority,
    double? progress,
    int? taskCount,
    int? completedTaskCount,
  }) {
    return BotWork(
      id: id,
      botId: botId,
      chatId: chatId,
      title: title ?? this.title,
      description: description ?? this.description,
      goal: goal ?? this.goal,
      functionId: functionId,
      scheduleId: scheduleId,
      parentWorkId: parentWorkId,
      status: status ?? this.status,
      priority: priority ?? this.priority,
      progress: progress ?? this.progress,
      createdAt: createdAt,
      startedAt: startedAt,
      completedAt: completedAt,
      dueAt: dueAt,
      result: result,
      error: error,
      context: context,
      tags: tags,
      taskCount: taskCount ?? this.taskCount,
      completedTaskCount: completedTaskCount ?? this.completedTaskCount,
    );
  }

  bool get isActive =>
      status == WorkStatus.pending || status == WorkStatus.inProgress;
  bool get isDone => status == WorkStatus.completed;
  bool get hasFailed => status == WorkStatus.failed;
}

class WorkTask {
  const WorkTask({
    required this.id,
    required this.botId,
    required this.workId,
    this.chatId,
    required this.title,
    this.description,
    this.action,
    this.order = 0,
    this.dependsOn = const [],
    required this.status,
    required this.priority,
    this.retryCount = 0,
    this.maxRetries = 0,
    this.timeoutSeconds,
    required this.createdAt,
    this.startedAt,
    this.completedAt,
    this.result,
    this.error,
  });

  final String id;
  final String botId;
  final String workId;
  final String? chatId;
  final String title;
  final String? description;
  final String? action;
  final int order;
  final List<String> dependsOn;
  final TaskStatus status;
  final Priority priority;
  final int retryCount;
  final int maxRetries;
  final int? timeoutSeconds;
  final DateTime createdAt;
  final DateTime? startedAt;
  final DateTime? completedAt;
  final String? result;
  final String? error;

  factory WorkTask.fromJson(Map<String, dynamic> json) {
    return WorkTask(
      id: json['id'] as String,
      botId: json['botId'] as String,
      workId: json['workId'] as String,
      chatId: json['chatId'] as String?,
      title: json['title'] as String,
      description: json['description'] as String?,
      action: json['action'] as String?,
      order: json['order'] as int? ?? 0,
      dependsOn: (json['dependsOn'] as List<dynamic>?)?.cast<String>() ?? [],
      status: TaskStatus.fromString(json['status'] as String),
      priority: Priority.fromString(json['priority'] as String? ?? 'normal'),
      retryCount: json['retryCount'] as int? ?? 0,
      maxRetries: json['maxRetries'] as int? ?? 0,
      timeoutSeconds: json['timeoutSeconds'] as int?,
      createdAt: DateTime.parse(json['createdAt'] as String),
      startedAt: json['startedAt'] != null
          ? DateTime.parse(json['startedAt'] as String)
          : null,
      completedAt: json['completedAt'] != null
          ? DateTime.parse(json['completedAt'] as String)
          : null,
      result: json['result'] as String?,
      error: json['error'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'botId': botId,
        'workId': workId,
        'title': title,
        'description': description,
        'action': action,
        'order': order,
        'dependsOn': dependsOn,
        'status': status.toApi(),
        'priority': priority.name,
        'createdAt': createdAt.toIso8601String(),
      };

  WorkTask copyWith({
    String? title,
    String? description,
    TaskStatus? status,
    Priority? priority,
    String? result,
    String? error,
  }) {
    return WorkTask(
      id: id,
      botId: botId,
      workId: workId,
      chatId: chatId,
      title: title ?? this.title,
      description: description ?? this.description,
      action: action,
      order: order,
      dependsOn: dependsOn,
      status: status ?? this.status,
      priority: priority ?? this.priority,
      retryCount: retryCount,
      maxRetries: maxRetries,
      timeoutSeconds: timeoutSeconds,
      createdAt: createdAt,
      startedAt: startedAt,
      completedAt: completedAt,
      result: result ?? this.result,
      error: error ?? this.error,
    );
  }

  bool get isRunning => status == TaskStatus.inProgress;
  bool get isDone => status == TaskStatus.completed;
  bool get hasFailed => status == TaskStatus.failed;
}

class Job {
  const Job({
    required this.id,
    required this.botId,
    required this.taskId,
    required this.workId,
    this.chatId,
    required this.status,
    required this.attempt,
    required this.progress,
    required this.createdAt,
    this.startedAt,
    this.completedAt,
    this.result,
    this.error,
    this.logs = const [],
  });

  final String id;
  final String botId;
  final String taskId;
  final String workId;
  final String? chatId;
  final JobStatus status;
  final int attempt;
  final double progress;
  final DateTime createdAt;
  final DateTime? startedAt;
  final DateTime? completedAt;
  final String? result;
  final String? error;
  final List<JobLog> logs;

  factory Job.fromJson(Map<String, dynamic> json) {
    return Job(
      id: json['id'] as String,
      botId: json['botId'] as String,
      taskId: json['taskId'] as String,
      workId: json['workId'] as String,
      chatId: json['chatId'] as String?,
      status: JobStatus.fromString(json['status'] as String),
      attempt: json['attempt'] as int? ?? 1,
      progress: (json['progress'] as num?)?.toDouble() ?? 0.0,
      createdAt: DateTime.parse(json['createdAt'] as String),
      startedAt: json['startedAt'] != null
          ? DateTime.parse(json['startedAt'] as String)
          : null,
      completedAt: json['completedAt'] != null
          ? DateTime.parse(json['completedAt'] as String)
          : null,
      result: json['result'] as String?,
      error: json['error'] as String?,
      logs: (json['logs'] as List<dynamic>?)
              ?.map((e) => JobLog.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }
}

class JobLog {
  const JobLog({
    required this.timestamp,
    required this.level,
    required this.message,
    this.data,
  });

  final DateTime timestamp;
  final String level;
  final String message;
  final Map<String, dynamic>? data;

  factory JobLog.fromJson(Map<String, dynamic> json) {
    return JobLog(
      timestamp: DateTime.parse(json['timestamp'] as String),
      level: json['level'] as String,
      message: json['message'] as String,
      data: json['data'] as Map<String, dynamic>?,
    );
  }
}

class Todo {
  const Todo({
    required this.id,
    required this.botId,
    this.chatId,
    required this.title,
    this.notes,
    required this.status,
    required this.priority,
    required this.createdAt,
    this.completedAt,
    this.remindAt,
    this.convertedToWorkId,
    this.convertedToTaskId,
    this.tags = const [],
  });

  final String id;
  final String botId;
  final String? chatId;
  final String title;
  final String? notes;
  final TodoStatus status;
  final Priority priority;
  final DateTime createdAt;
  final DateTime? completedAt;
  final DateTime? remindAt;
  final String? convertedToWorkId;
  final String? convertedToTaskId;
  final List<String> tags;

  factory Todo.fromJson(Map<String, dynamic> json) {
    return Todo(
      id: json['id'] as String,
      botId: json['botId'] as String,
      chatId: json['chatId'] as String?,
      title: json['title'] as String,
      notes: json['notes'] as String?,
      status: TodoStatus.fromString(json['status'] as String),
      priority: Priority.fromString(json['priority'] as String? ?? 'normal'),
      createdAt: DateTime.parse(json['createdAt'] as String),
      completedAt: json['completedAt'] != null
          ? DateTime.parse(json['completedAt'] as String)
          : null,
      remindAt: json['remindAt'] != null
          ? DateTime.parse(json['remindAt'] as String)
          : null,
      convertedToWorkId: json['convertedToWorkId'] as String?,
      convertedToTaskId: json['convertedToTaskId'] as String?,
      tags: (json['tags'] as List<dynamic>?)?.cast<String>() ?? [],
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'botId': botId,
        'chatId': chatId,
        'title': title,
        'notes': notes,
        'status': status.name,
        'priority': priority.name,
        'createdAt': createdAt.toIso8601String(),
        'completedAt': completedAt?.toIso8601String(),
        'remindAt': remindAt?.toIso8601String(),
        'tags': tags,
      };

  Todo copyWith({
    String? title,
    String? notes,
    TodoStatus? status,
    Priority? priority,
    DateTime? remindAt,
    List<String>? tags,
  }) {
    return Todo(
      id: id,
      botId: botId,
      chatId: chatId,
      title: title ?? this.title,
      notes: notes ?? this.notes,
      status: status ?? this.status,
      priority: priority ?? this.priority,
      createdAt: createdAt,
      completedAt: completedAt,
      remindAt: remindAt ?? this.remindAt,
      convertedToWorkId: convertedToWorkId,
      convertedToTaskId: convertedToTaskId,
      tags: tags ?? this.tags,
    );
  }

  bool get isOpen => status == TodoStatus.open;
  bool get isDone => status == TodoStatus.done;
  bool get isDismissed => status == TodoStatus.dismissed;
}
